File: pkg/planner/core/operator/physicalop/physical_union_all.go
// Copyright 2025 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package physicalop

import (
	"github.com/pingcap/tidb/pkg/planner/core/base"
	"github.com/pingcap/tidb/pkg/planner/core/operator/logicalop"
	"github.com/pingcap/tidb/pkg/planner/property"
	"github.com/pingcap/tidb/pkg/planner/util/costusage"
	"github.com/pingcap/tidb/pkg/planner/util/utilfuncp"
	"github.com/pingcap/tidb/pkg/util/plancodec"
	"github.com/pingcap/tidb/pkg/util/size"
)

// PhysicalUnionAll is the physical operator of UnionAll.
type PhysicalUnionAll struct {
	PhysicalSchemaProducer

	Mpp bool
}

// Init initializes PhysicalUnionAll.
func (p PhysicalUnionAll) Init(ctx base.PlanContext, stats *property.StatsInfo, offset int, props ...*property.PhysicalProperty) *PhysicalUnionAll {
	p.BasePhysicalPlan = NewBasePhysicalPlan(ctx, plancodec.TypeUnion, &p, offset)
	p.SetChildrenReqProps(props)
	p.SetStats(stats)
	return &p
}

// Clone implements op.PhysicalPlan interface.
func (p *PhysicalUnionAll) Clone(newCtx base.PlanContext) (base.PhysicalPlan, error) {
	cloned := new(PhysicalUnionAll)
	cloned.SetSCtx(newCtx)
	base, err := p.PhysicalSchemaProducer.CloneWithSelf(newCtx, cloned)
	if err != nil {
		return nil, err
	}
	cloned.PhysicalSchemaProducer = *base
	return cloned, nil
}

// MemoryUsage return the memory usage of PhysicalUnionAll
func (p *PhysicalUnionAll) MemoryUsage() (sum int64) {
	if p == nil {
		return
	}

	return p.PhysicalSchemaProducer.MemoryUsage() + size.SizeOfBool
}

// Attach2Task implements base.PhysicalPlan interface.
func (p *PhysicalUnionAll) Attach2Task(tasks ...base.Task) base.Task {
	return utilfuncp.Attach2Task4PhysicalUnionAll(p, tasks...)
}

// GetPlanCostVer1 implements base.PhysicalPlan interface.
func (p *PhysicalUnionAll) GetPlanCostVer1(taskType property.TaskType, option *costusage.PlanCostOption) (float64, error) {
	return utilfuncp.GetPlanCostVer14PhysicalUnionAll(p, taskType, option)
}

// GetPlanCostVer2 implements base.PhysicalPlan interface.
func (p *PhysicalUnionAll) GetPlanCostVer2(taskType property.TaskType, option *costusage.PlanCostOption, _ ...bool) (costusage.CostVer2, error) {
	return utilfuncp.GetPlanCostVer24PhysicalUnionAll(p, taskType, option)
}

// ExhaustPhysicalPlans4LogicalUnionAll generates PhysicalUnionAll plans from LogicalUnionAll.
func ExhaustPhysicalPlans4LogicalUnionAll(p *logicalop.LogicalUnionAll, prop *property.PhysicalProperty) ([]base.PhysicalPlan, bool, error) {
	// TODO: UnionAll can not pass any order, but we can change it to sort merge to keep order.
	if !prop.IsSortItemEmpty() || (prop.IsFlashProp() && prop.TaskTp != property.MppTaskType) {
		return nil, true, nil
	}
	// TODO: UnionAll can pass partition info, but for briefness, we prevent it from pushing down.
	if prop.TaskTp == property.MppTaskType && prop.MPPPartitionTp != property.AnyType {
		return nil, true, nil
	}
	// when arrived here, operator itself has already checked checkOpSelfSatisfyPropTaskTypeRequirement, we only need to feel allowMPP here.
	canUseMpp := p.SCtx().GetSessionVars().IsMPPAllowed()
	chReqProps := make([]*property.PhysicalProperty, 0, p.ChildLen())
	for range p.Children() {
		if canUseMpp && prop.TaskTp == property.MppTaskType {
			chReqProps = append(chReqProps, &property.PhysicalProperty{
				ExpectedCnt:       prop.ExpectedCnt,
				TaskTp:            property.MppTaskType,
				CTEProducerStatus: prop.CTEProducerStatus,
				NoCopPushDown:     prop.NoCopPushDown,
			})
		} else {
			chReqProps = append(chReqProps, &property.PhysicalProperty{ExpectedCnt: prop.ExpectedCnt,
				CTEProducerStatus: prop.CTEProducerStatus, NoCopPushDown: prop.NoCopPushDown})
		}
	}
	ua := PhysicalUnionAll{
		Mpp: canUseMpp && prop.TaskTp == property.MppTaskType,
	}.Init(p.SCtx(), p.StatsInfo().ScaleByExpectCnt(p.SCtx().GetSessionVars(), prop.ExpectedCnt), p.QueryBlockOffset(), chReqProps...)
	ua.SetSchema(p.Schema())
	if canUseMpp && prop.TaskTp == property.RootTaskType {
		chReqProps = make([]*property.PhysicalProperty, 0, p.ChildLen())
		for range p.Children() {
			chReqProps = append(chReqProps, &property.PhysicalProperty{
				ExpectedCnt:       prop.ExpectedCnt,
				TaskTp:            property.MppTaskType,
				CTEProducerStatus: prop.CTEProducerStatus,
				NoCopPushDown:     prop.NoCopPushDown,
			})
		}
		mppUA := PhysicalUnionAll{Mpp: true}.Init(p.SCtx(), p.StatsInfo().ScaleByExpectCnt(p.SCtx().GetSessionVars(), prop.ExpectedCnt), p.QueryBlockOffset(), chReqProps...)
		mppUA.SetSchema(p.Schema())
		return []base.PhysicalPlan{ua, mppUA}, true, nil
	}
	return []base.PhysicalPlan{ua}, true, nil
}

// ExhaustPhysicalPlans4LogicalPartitionUnionAll generates PhysicalUnionAll plans from LogicalPartitionUnionAll.
func ExhaustPhysicalPlans4LogicalPartitionUnionAll(p *logicalop.LogicalPartitionUnionAll, prop *property.PhysicalProperty) ([]base.PhysicalPlan, bool, error) {
	uas, flagHint, err := ExhaustPhysicalPlans4LogicalUnionAll(&p.LogicalUnionAll, prop)
	if err != nil {
		return nil, false, err
	}
	for _, ua := range uas {
		ua.(*PhysicalUnionAll).SetTP(plancodec.TypePartitionUnion)
	}
	return uas, flagHint, nil
}

File: rust/difftests/planner-tests/tests/physical_union_all.rs
// Copyright 2026 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Vectors for the wired PhysicalUnionAll planning gates.
//!
//! The Go anchor is `TestMppUnionAll` at
//! `pkg/planner/core/casetest/mpp/mpp_test.go:446`.

use tidb_planner::logical::{BaseLogicalPlan, LogicalUnionAll};
use tidb_planner::physical::{exhaust_physical_plans_4_logical_union_all, PhysicalPlan};
use tidb_planner::physical_property::{PhysicalProperty, SortItem, TaskType};
use tidb_planner::plan_base::PlanIdAllocator;

#[test]
fn unsupported_mpp_union_is_not_claimed_without_the_tiflash_tier() {
    let allocator = PlanIdAllocator::new();
    let union = LogicalUnionAll::new(BaseLogicalPlan::new(&allocator, LogicalUnionAll::TYPE, 0));
    let mpp = PhysicalProperty {
        task_tp: TaskType::Mpp,
        ..PhysicalProperty::default()
    };
    assert!(exhaust_physical_plans_4_logical_union_all(&union, &mpp, &allocator, 1.0).is_empty());
}

#[test]
fn root_union_all_emits_source_candidate_order() {
    let allocator = PlanIdAllocator::new();
    let union = LogicalUnionAll::new(BaseLogicalPlan::new(&allocator, LogicalUnionAll::TYPE, 7));
    let plans = exhaust_physical_plans_4_logical_union_all(
        &union,
        &PhysicalProperty::default(),
        &allocator,
        1.0,
    );
    assert_eq!(
        plans.len(),
        1,
        "the TiFlash-less tier has one root candidate"
    );
    let PhysicalPlan::UnionAll(plan) = &plans[0] else {
        panic!("a wired PhysicalUnionAll, got {:?}", plans[0]);
    };
    assert_eq!(plan.base.base.tp(), "Union");
    assert_eq!(plan.base.base.query_block_offset(), 7);
    assert!(!plan.mpp);
}

#[test]
fn sort_requests_are_rejected() {
    let allocator = PlanIdAllocator::new();
    let union = LogicalUnionAll::new(BaseLogicalPlan::new(&allocator, LogicalUnionAll::TYPE, 0));
    let sorted = PhysicalProperty {
        sort_items: vec![SortItem::new(1, false)],
        ..PhysicalProperty::default()
    };
    assert!(
        exhaust_physical_plans_4_logical_union_all(&union, &sorted, &allocator, 1.0).is_empty()
    );
}
